### PR TITLE
Add ability to change Base fee manually 

### PR DESCRIPTION
**Pull request by 11Yupi11**
**Summary**

Special transaction to change the base fee was created and some restrictions about who can do it and in which ammount it is possible to do.

**Purpose of change** 

For some period of time Camino token will be stable, but due to its changes in future we need to have mechanism to change also base fee ammount not to make it overprised or too cheap. The idea is to keep it in amount of 1 cent price. 

**Solution:** 

- SunrisePhase0 constant of base fee was deleted
- Created C4Taddress as header of block in genesis file
- To stateDB new StateObject was added on the specific address (the same as C4Taddress, as it is stored in every block and easy to get it in all parts of the transaction)
- Added function to read information from stateDB
- Added function to change StateObject to the new value
- To change the Base Fee you need to send transaction from C-chain address of C4T account to the same address and set in value the ammount of BaseFee you want in nanoCAM, but for now it might not be more than 0.000000225

**What need to be done:**

- Clean the code and make it better
- Fix the tests (linters failed also due to errors in tests - not enough arguments in call he function)
NOTE: block_creation tests might be fixed after all modifications to genesis file will be done

### CODE REVIEW

File: consensus/consensus.go [hunk 1]
```diff
@@ -64,6 +64,9 @@ type ChainHeaderReader interface {
 
 	// GetHeaderByHash retrieves a block header from the database by its hash.
 	GetHeaderByHash(hash common.Hash) *types.Header
+
```
Comment (DerTiedemann) [inline consensus/consensus.go:68]: Id recommend to extract this into an seperate interface instead of using the existing one.
Just for clarity.

**Comment by DerTiedemann** [comment]
I would love to have the basefee change functionality and the accumulative address to be in seperate PR so its easier to write test for them. Also we should figure out what test we need to fix and which we can ignore for now.

File: core/state/state_object.go [hunk 2]
```diff
@@ -129,6 +129,10 @@ func newObject(db *StateDB, address common.Address, data types.StateAccount) *st
 	if data.Balance == nil {
 		data.Balance = new(big.Int)
 	}
+	if data.BaseFee == nil {
+		otherBaseFee := db.GetBaseFee(common.HexToAddress("0x36d86bb280b4e088793a60c781c7edcd9fa33255"))
```
Comment (DerTiedemann) [inline core/state/state_object.go]: This needs to be extracted to a seperate address in genesis or just use the accumualtive address.
If so, lets rename it to root or superuser address.

File: core/state/statedb.go [hunk 3]
```diff
@@ -332,6 +332,19 @@ func (s *StateDB) GetCode(addr common.Address) []byte {
 	return nil
 }
 
+func (s *StateDB) GetBaseFee(addr common.Address) *big.Int {
+	stateObject := s.getStateObject(addr)
+	if stateObject != nil && stateObject.BaseFee() != nil {
+		return stateObject.BaseFee()
+	}
+	return big.NewInt(225_000_000_000)
```
Comment (DerTiedemann) [inline core/state/statedb.go]: This needs to be configurable e.g. in genesis or use the baseline sunrise phae value

File: core/state_transition.go [hunk 4]
```diff
@@ -350,7 +350,20 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
 	st.refundGas(apricotPhase1)
-	st.state.AddBalance(st.evm.Context.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.gasPrice))
+
+	totalFee := new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.gasPrice)
+	onePercentCalc := new(big.Int).Div(totalFee, new(big.Int).SetUint64(100))
+	exportedPartFee := new(big.Int).Mul(onePercentCalc, new(big.Int).SetUint64(60))
+	burnedPartFee := new(big.Int).Sub(totalFee, exportedPartFee)
+	st.state.AddBalance(st.evm.Context.AccumulativeAddress, exportedPartFee)
+	st.state.AddBalance(st.evm.Context.Coinbase, burnedPartFee)
+	if st.msg.From() == common.HexToAddress("0xc02104f25e07f827195cebf032a1d7bfecf0cb7c") && *st.msg.To() == common.HexToAddress("0xc02104f25e07f827195cebf032a1d7bfecf0cb7c") {
```
Comment (DerTiedemann) [inline core/state_transition.go]: Use configurable address here.

**Review by DerTiedemann — CHANGES_REQUESTED** [review]
This pr is a little broken and needs a lot of cleanup.
Id suggest we split them into two and work on the two features seperatly.
E.g. one to add a superuser/accumulative address we can use to save the part of the basefee to send to validators. And another one to facilitate the basefee change.

File: core/state/state_object.go (hunk 2 above)
Comment (11Yupi11) [reply core/state/state_object.go]: It is Accumulative address.....i can't get it from this directory

File: core/state_transition.go (hunk 4 above)
Comment (11Yupi11) [reply core/state_transition.go]: I want to ask if it is a good idea just to add specific address like "C4Taddress" to genesis file to use it.

**Comment by DerTiedemann** [comment]
Can you have a look at the failing build and have a go at fixing it, seems like some parts were not updated to som function siganture check

File: core/vm/evm.go [hunk 5]
```diff
@@ -98,6 +98,7 @@ type BlockContext struct {
 
 	// Block information
 	Coinbase    common.Address // Provides information for COINBASE
+	C4Taddress  common.Address // Provides information for C4Taddress
```
Comment (c4t-ag) [inline core/vm/evm.go:101]: Why do we need a new field in such a general structure?

File: core/state_transition.go [hunk 6]
```diff
@@ -350,8 +350,17 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 		ret, st.gas, vmerr = st.evm.Call(sender, st.to(), st.data, st.gas, st.value)
 	}
 	st.refundGas(apricotPhase1)
+
 	st.state.AddBalance(st.evm.Context.Coinbase, new(big.Int).Mul(new(big.Int).SetUint64(st.gasUsed()), st.gasPrice))
 
+	if st.msg.From() == st.evm.Context.C4Taddress && *st.msg.To() == st.evm.Context.C4Taddress {
```
Comment (c4t-ag) [inline core/state_transition.go:356]: For me this feels quite hacky.
Beside this To() can return nil if I'm not wrong which will kill e.g. contract creations.

File: core/state_transition.go [hunk 7]
```diff
@@ -350,8 +350,17 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
...
+		if st.msg.Value().Cmp(big.NewInt(params.ApricotPhase3InitialBaseFee)) <= 0 {
```
Comment (c4t-ag) [inline core/state_transition.go:357]: Why do we have to deal with ApricotPhase3 here?

File: core/tx_pool.go [hunk 8]
```diff
@@ -1715,7 +1715,7 @@ func (pool *TxPool) updateBaseFee() {
 	pool.mu.Lock()
 	defer pool.mu.Unlock()
 
-	_, baseFeeEstimate, err := dummy.EstimateNextBaseFee(pool.chainconfig, pool.currentHead, uint64(time.Now().Unix()))
+	_, baseFeeEstimate, err := dummy.EstimateNextBaseFee(pool.chainconfig, pool.currentHead, pool.currentState, uint64(time.Now().Unix()))
```
Comment (c4t-ag) [inline core/tx_pool.go:1718]: Is it really required to pass state here? In estimate... we already query state if I'm not wrong.

File: eth/gasprice/gasprice.go [hunk 9]
```diff
@@ -357,7 +357,8 @@ func (oracle *Oracle) suggestDynamicFees(ctx context.Context) (*big.Int, *big.In
 	}
 
 	if oracle.backend.ChainConfig().IsSunrisePhase0(new(big.Int).SetUint64(head.Time)) {
-		baseFee = new(big.Int).SetUint64(params.SunrisePhase0BaseFee)
```
Comment (c4t-ag) [inline eth/gasprice/gasprice.go:360]: This was introduced to make MetaMask happy, why is it reverted?

File: consensus/dummy/dynamic_fees.go [hunk 10]
```diff
@@ -69,9 +70,8 @@ func CalcBaseFee(config *params.ChainConfig, parent *types.Header, timestamp uin
 	if timestamp < parent.Time {
 		return nil, nil, fmt.Errorf("cannot calculate base fee for timestamp (%d) prior to parent timestamp (%d)", timestamp, parent.Time)
 	}
-
 	if isSunrisePhase0 {
-		return []byte{}, new(big.Int).SetUint64(params.SunrisePhase0BaseFee), nil
+		return []byte{}, parent.BaseFee, nil
```
Comment (c4t-ag) [inline consensus/dummy/dynamic_fees.go:74]: This kills the Sunrise definition, that changes should apply earliest after the first block in the new Phase.

File: core/types/state_account.go [hunk 11]
```diff
@@ -37,6 +37,7 @@ import (
 type StateAccount struct {
 	Nonce       uint64
 	Balance     *big.Int
+	BaseFee     *big.Int
```
Comment (c4t-ag) [inline core/types/state_account.go:40]: If I'm not wrong this is the wrong place for a "BaseFee"

**Review by c4t-ag — CHANGES_REQUESTED** [review]
This PR needs a discussion about the concept and clarifications about change of existing behaviour. We should look for a solution with less invasive changes b.t.w.
Beside this this PR has to be cleaned up, and CI (linter issues) should be fixed, but pls. wait with that until we discussed the other topics.

**Comment by peak3d** [comment]
Replaced by #15

**Pull request closed without merging**